### PR TITLE
New Extension: Break Apart and Retain Holes

### PR DESCRIPTION
**Pull request by kaalleen**
This aims to be a new extension as suggested in #360 
It's located in the submenu "Fill Tools" (because it only makes sense for fill areas).

### CODE REVIEW

**Comment by lexelby** [comment]
This is the best idea ever.  I've been thinking about a way to handle fill stitch for paths that contain separate unconnected shapes, but I hadn't thought through how to handle the holes.  If this works we can use this code for that part!

**Comment by kaalleen** [comment]
The original idea was yours ;)

It'd be great if it could be of use for handling unconnected shapes without throwing an error message. For those who want to separated the elements with only one command (maybe to order them precisely), it'd be still useful as an extension.

I don't know if I maybe used too many for loops. Python has nice tools to avoid them, but I might not be aware of them :D

File: lib/extensions/break_apart.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                continue
+
+            polygons = []
+            multipolygons = []
+            holes = []
+
+            # sort paths by size and convert to polygons
+            element.paths.sort(key=lambda point_list: shgeo.Polygon(point_list).area, reverse=True)
+            for path in element.paths:
+                polygons.append(shgeo.Polygon(path))
```
Comment (lexelby) [inline lib/extensions/break_apart.py:35]: How about sort `polygons` after the loop?  That way we don't have to convert each path to a `shgeo.Polygon` twice.

File: lib/extensions/break_apart.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                multipolygon = [shape]
+
+                for other in polygons:
+                    if shape != other and shape.contains(other) and other not in holes:
+                        # check if "other" is inside a hole, before we add it to the list
+                        if any(p.contains(other) for p in multipolygon[1:]):
+                            continue
+                        multipolygon.append(other)
+                        holes.append(other)
+                multipolygons.append(shgeo.MultiPolygon(multipolygon))
```
Comment (lexelby) [inline lib/extensions/break_apart.py:49]: I don't think we have to use `MultiPolygon` here.  A Shapely `Polygon` has an outer boundary and multiple holes.  A `MultiPolygon` is a container that can contain multiple `Polygon` instances.

**Comment by lexelby** [comment]
I don't think you have too many loops!  It only has two nested loops and I think it's probably gonna run fast even for complicated shapes.

I thought of a case that I think might cause trouble.  What about a "bullseye" shape, something like this:

bullseye

We don't want the outer circle to "eat" all of the middle circles.  Maybe we should make it skip adding an `other` if the `other` is contained by one of the existing holes?

**Comment by lexelby** [comment]
OMG never mind, your code already does that!  Awesome :D :D

**Comment by kaalleen** [comment]
Thank you for your review. These changes totally make sense. I've updated this branch.
The "bullseye" problem really seems to work great now :)
I have spend some time to think about it though.

**Pull request merged**